esp8266-fastled-webserver/Fields.cpp

```diff
@@ -17,479 +17,526 @@
-}
-String setSHueMin(String value) {
-  sHueMin = value.toInt();
-  return value;
-}
+    String getSHueMax() {
+      return String(sHueMax);
+    }
+    String setSHueMax(String value) {
+      sHueMax = value.toInt();
+      return value;
+    }
 
-String getSHueMax() {
-  return String(sHueMax);
-}
-String setSHueMax(String value) {
-  sHueMax = value.toInt();
-  return value;
-}
+  }
 
+  inline namespace Options {
+    void getPatterns(JsonVariant dst) {
+      for (uint8_t i = 0; i < patternCount; i++) {
+        dst.add(patterns[i].name);
+      }
+    }
 
-typedef String (*FieldSetter)(String);
-typedef String (*FieldGetter)();
-struct Field {
-  String name;
-  String label;
-  String type;
-  uint8_t min;
-  uint8_t max;
-  FieldGetter getValue;
-  FieldGetter getOptions;
-  FieldSetter setValue;
-};
-
-// passing array reference works fine, but need to make the function a template
-// to capture the array size... on the positive side, no need to pass `count` parameter
-template <size_t N>
-Field  getField(String name, const Field (&fields)[N]) {
-  for (uint8_t i = 0; i < N; i++) {
-    Field field = fields[i];
-    if (field.name == name) {
-      return field;
+    void getPalettes(JsonVariant dst) {
+      for (uint8_t i = 0; i < paletteCount; i++) {
+        dst.add(paletteNames[i]);
+      }
     }
   }
-  return Field();
-}
-template <size_t N>
-String getFieldValue(String name, const Field (&fields)[N]) {
-  Field field = getField(name, fields);
-  if (field.getValue) {
-    return field.getValue();
-  }
-  return String();
-}
-template <size_t N>
-String setFieldValue(String name, String value, const Field (&fields)[N]) {
-  Field field = getField(name, fields);
-  if (field.setValue) {
-    return field.setValue(value);
-  }
-  return String();
-}
-template <size_t N>
-String getFieldsJson(const Field (&fields)[N]) {
-  String json = "[";
-
-  for (uint8_t i = 0; i < N; i++) {
-    Field field = fields[i];
 
-    json += "{\"name\":\"" + field.name + "\",\"label\":\"" + field.label + "\",\"type\":\"" + field.type + "\"";
+  typedef String (*FieldSetter)(String);
+  typedef String (*FieldGetter)();
+  typedef void (*FieldOptions)(JsonVariant);
+  struct Field {
+    const String name;
+    const String label;
+    Field_t type;
+    uint8_t min;
+    uint8_t max;
+    FieldGetter  getValue;
+    FieldOptions getOptions;
+    FieldSetter  setValue;
+  };
+
+  void convertToJson(const Field& field, JsonVariant dst) {
+    if (field.name.length() == 0) {
+      return;
+    }
+    if (!IsValid(field.type)) {
+      return;
+    }
 
-    if(field.getValue) {
-      if (field.type == ColorFieldType || field.type == StringFieldType || field.type == LabelFieldType) {
-        json += ",\"value\":\"" + field.getValue() + "\"";
-      }
-      else {
-        json += ",\"value\":" + field.getValue();
+    dst[F("name")]  = field.name;
+    dst[F("label")] = field.label;
+
+    dst[F("type")]  = ToString(field.type);
```

Comment:
If the modification to use enumeration in app.js as well is done, this needs to be change to:
`dst[F("type")]  = (uint8_t)field.type;`
```suggestion
    dst[F("type")]  = (uint8_t)field.type;
```

Reply:
Yes, if this was to be done.  Currently, I am trying to keep compatibility, in case folks have relied upon this JSON output (e.g., using their own parser) to display the state of the webserver.  If I were to change the type, this would break anything existing code expecting the stringified version.  Thus, I decided against any breaking changes at this time.

Hope that makes sense?

Reply:
For sure certainly does make sense.
I just doubt that there are actually many (or any) people relying on the "/all" output apart from the web app itself.
I do but I also completely changed the project for my personal taste and use....

And by introducing a utc offset  as additional field with explicit values being required on the web app, compatibility is somewha broken anyway ;-)